### PR TITLE
(WIP) Vertical scrolling page to test overflow:auto

### PR DESCRIPTION
**Pull request by BrianSipple**
This is still very early on, but so far it represents my idea for addressing https://github.com/ember-a11y/ember-a11y/issues/9.

I'm envisioning something like a `friends` route where one of the child routes (the only one, so far) is a [list of suggested friends](https://github.com/BrianSipple/ember-a11y/blob/vertical-scrolling-find-friends-page/tests/dummy/app/routes/friends/find.js) where [the `focusing-outlet` is inside of a container](https://github.com/BrianSipple/ember-a11y/blob/vertical-scrolling-find-friends-page/tests/dummy/app/templates/friends.hbs) with `overflow: auto` and has a [fair amount of content to scroll down through](https://github.com/BrianSipple/ember-a11y/blob/vertical-scrolling-find-friends-page/tests/dummy/app/templates/friends/find.hbs)

At the moment, though, I still have to dig in to the mechanics of how the [outlet's `setFocus` method treats this](https://github.com/nathanhammond/ember-a11y/blob/master/app/components/focusing-outlet.js#L41).

### CODE REVIEW

**Comment by BrianSipple** [comment]
So here's the what I've been able to find so far: When we're calling `this.element.focus()` on an outlet, the browser will scroll to it. 

If I'm understanding #9 correctly and the goal is to prevent this -- which is what I've been attempting -- [calling `window.scrollTo(scrollLeft, scrollTop)`](https://github.com/ember-a11y/ember-a11y/pull/30/files#diff-1c8dddccb1c8f809b7da1fe7a89b66bcR56) just after `this.element.setFocus()`, within the same run-loop function, should do the trick. 

But it doesn't. 

Are we attempting to do anything else with scroll behavior at the same time somewhere else that I'm just not seeing?

**Comment by nathanhammond** [comment]
We already support not scrolling to an outlet, the only problem right now is if that outlet is inside of a block with `overflow: scroll;` the behavior is `undefined`. I don't know what we want our behavior to be, but we should definitely create a test for it, and then determine what our proper reaction is. (My suspicion is that we don't want to trigger any visual scrolling, so resetting all positions.)

The issue you're encountering is that all scroll handling has to occur _after_ the scroll event has fired, which is what we are doing here:
https://github.com/ember-a11y/ember-a11y/blob/master/app/components/focusing-outlet.js#L9

Sync in the same loop hasn't yet yielded the JS stack frame so that the `scroll` handler can fire. In order to do this we need to grab all of the offset positions, save them, and then after the event has fired, reset all of them in the `handler` function. This probably requires walking the full set of `.parentNode` of the outlet and see if any of them could be scrollable.

**Comment by nathanhammond** [comment]
Flow:
1. https://github.com/ember-a11y/ember-a11y/blob/master/app/components/focusing-outlet.js#L47
2. https://github.com/ember-a11y/ember-a11y/blob/master/app/components/focusing-outlet.js#L50
3. https://github.com/ember-a11y/ember-a11y/blob/master/app/components/focusing-outlet.js#L9-L10

The `scroll` event actually has to fire, we let the browser move you to that position, but immediately set it back. You can't actually cancel a scroll event, so we're just resetting the consequences of that event.

**Comment by BrianSipple** [comment]
@nathanhammond Gotcha. I was debugging both areas and it doesn't seem that the handler is actually being called. Only when I listen [here](https://github.com/ember-a11y/ember-a11y/blob/master/app/components/focusing-outlet.js#L47) to the `focus` event that gets fired when calling `this.element.focus()` does it get reached. 

With that in mind, I'm starting to think it might be best to track a standalone issue for testing/determining how we handle scrolling on focus for different types of overflow settings -- as I'm seeing a number of ways (you hinted at one with walking `.parentNode`) that the `focusing-outlet` can be improved to help with whatever our desired behavior becomes. 

If that sounds okay, I could make this PR only include the changes related to setting up the new page in the first place (and perhaps listening to `onfocus`), and then get another WIP up for the new issue.

File: app/components/focusing-outlet.js [hunk 1]
```diff
@@ -42,18 +47,32 @@ let FocusingOutlet = Ember.Component.extend({
...
+      // focusing outlet's container?
       scrollLeft = document.body.scrollLeft;
       scrollTop = document.body.scrollTop;
-      window.addEventListener('scroll', handler);
+
+      window.addEventListener('focus', focusHandler);
 
       // Set the focus to the target outlet wrapper.
-      Ember.run.schedule('afterRender', this, function() { this.element.focus(); });
+      Ember.run.scheduleOnce('afterRender', () => {
```
Comment (nathanhammond) [inline app/components/focusing-outlet.js]: So, `scheduleOnce` requires a named function to test for equivalence.

**Comment by nathanhammond** [comment]
If you resize the scroll area of the present demo application it doesn't actually trigger a scroll event on window: the focus target is in an `overflow: scroll;` container. Whoops. 😜

[If you go back in time, this is an example of the original approach working.](https://github.com/ember-a11y/ember-a11y/tree/f11b0584b8c76fb3cd822743cf8f48813720ea82)

**Comment by BrianSipple** [comment]
@nathanhammond I'm not sure I follow. Resizing the scroll area as in resizing the window? Or is there something on resizable areas in this [demo app](https://ember-a11y.github.io/a11y-demo-app/#/)?

Also, regardless of whether or not I register the event listener on the element or on the window, and regardless of whether or not I call `this.element.focus()` synchronously or schedule it on the run loop's `afterRender` queue, the handler only fires when I listen for the `focus` event. (as opposed to `scroll`).

But then check this out: in the `overflow: auto` outlet (the friends list), calling `this.element.focus()` synchronously focuses the outlet, but doesn't trigger any browser-default scrolling behavior. However, calling it inside of a function scheduled in `afterRender` does cause the sudden snap to the top of the outlet -- in Chrome and Safari, but not Firefox. [I updated the branch to easily switch back and forth each approach]().

**Comment by nathanhammond** [comment]
I believe we're talking a bit past each other, including this message. Let's find some time to chat. 😄

```
$ git status
HEAD detached at f11b058
```
```diff
diff --git a/bower.json b/bower.json
index 8d0e8ce..29a382e 100644
--- a/bower.json
+++ b/bower.json
@@ -6,7 +6,9 @@
     "ember-cli-test-loader": "0.2.2",
     "ember-load-initializers": "0.1.7",
     "ember-qunit-notifications": "0.1.0",
-    "jquery": "2.1.4"
+    "jquery": "2.1.4",
+    "loader.js": "^3.0.0",
+    "qunit": "^1.23.1"
   },
   "resolutions": {
     "ember": "2.3.0"
```
Demo of scroll event firing.

Scroll event handling works at the `window` level if that's the only scroll event being triggered by the change. But you've made me notice something interesting: that handler should be unregistered in the case where the `scroll` event _doesn't_ fire because page height is smaller than the window.

This is not going to be easy. 😛

**Comment by nathanhammond** [comment]
@BrianSipple new thought: maybe we don't actually rely on `focus`/`scroll` events? Figure out what is being focused, walk `.parentNode` to the root, caching `scrollTop` sync. Use `setTimeout(walk, 0);` to get a new macro task, and reset all elements to their previous scroll position? Or maybe it would work in a micro task and we can `RSVP.resolve().then(walk);`. Or maybe we do it on `requestAnimationFrame`...

**Comment by BrianSipple** [comment]
@nathanhammond I think I’m getting there on how scroll events can be captured, but it still seems like listening to `focus`  -- and doing so on the element itself, not the window -- would be more inconsistent -- guaranteed, even. (And I’m still intrigued by how there doesn’t seem to be any scroll behavior at all when we call `this.element.focus()` synchronously.)

In any case, though, your mention of tearing down the handler sort of alludes to what I was getting at [here](https://github.com/ember-a11y/ember-a11y/pull/30/files#diff-1c8dddccb1c8f809b7da1fe7a89b66bcR83). When it comes to component event handlers, I typically initialize a handler function and assign it to the component during the `didInsertElement` hook, and then clean it up in `willDestroyElement`. I could PR that change separately if you’d like 😄 .

Also, [regarding the scroll handling idea](https://github.com/ember-a11y/ember-a11y/pull/30#issuecomment-224025802), I've been kicking something like that around in my head, too -- if we decided that we wanted to prevent scrolling. That said (and apologies if it's been discussed already), I’d be curious to hear what @ember-a11y/core thinks are the pros and cons of preventing vs not preventing. Part of me is worried that overriding the browser's default behavior could get us into trouble when we have several `focusing-outlets`, and, say, someone tries tabbing to an outlet off screen and expects the window scroll position to follow.

**Comment by nathanhammond** [comment]
@BrianSipple We are currently only overriding the behavior for imperatively triggered scroll resulting from calling `setFocus` on route transition. That seems fine. Already there are weird scrolling artifacts in Ember if you don't manage them on transition so we should probably work toward a defined behavior for that.

**Comment by MelSumner** [comment]
Due to the age of this PR, I'm going to close it- please reopen if there are new developments. Thank you!

**Pull request closed without merging**